### PR TITLE
add features and tests

### PR DESCRIPTION
**Pull request by bout3fiddy**

### CODE REVIEW

File: .github/workflows/unit-tests.yaml [hunk 1]
```diff
@@ -0,0 +1,30 @@
+name: unit-tests-boa
+
+on: ["push", "pull_request"]
```
Comment (DanielSchiavini) [inline .github/workflows/unit-tests.yaml:3]: just pick either push or pull_request, using both generates twice the same job for every commit in a PR

File: .github/workflows/unit-tests.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+      - uses: actions/checkout@v2
+
+      - name: Cache Compiler Installations
+        uses: actions/cache@v2
+        with:
+          path: |
+            ~/.vvm
+          key: compiler-cache
+
+      - name: Setup Python 3.10.4
```
Comment (DanielSchiavini) [inline .github/workflows/unit-tests.yaml:21]: Why not 3.11 or 3.12?

File: requirements.txt [hunk 3]
```diff
@@ -19,5 +19,5 @@ pdbpp
 hypothesis>=6.68.1
 
 # vyper and dev framework:
-git+https://github.com/vyperlang/titanoboa@40f5bfcc2afe212bb5a6f5026148f3625596ded7
+git+https://github.com/vyperlang/titanoboa@b5e9fb96d1424ed5cc5a6af03391d885439c83e5
```
Comment (DanielSchiavini) [inline requirements.txt:22]: It'd be nice to add a tag to the titanoboa commit

File: tests/utils/simulation_int_many.py [hunk 4]
```diff
@@ -0,0 +1,540 @@
...
+
+    def y(self, x, i, j):
+        xp = self.xp()
+        xp[i] = x * self.p[i] // 10**18
+        yp = solve_x(self.A, self.gamma, xp, self.D(), j)
+        return yp * 10**18 // self.p[j]
+
+
+def get_data(fname):
+    with open("download/{0}-1m.json".format(fname), "r") as f:
```
Comment (DanielSchiavini) [inline tests/utils/simulation_int_many.py:174]: Where's this coming from?

File: .github/workflows/unit-tests.yaml (hunk 1 above)
Comment (DanielSchiavini) [reply .github/workflows/unit-tests.yaml:3]: image

File: .github/workflows/unit-tests.yaml (hunk 2 above)
Comment (bout3fiddy) [reply .github/workflows/unit-tests.yaml:21]: That is irrelevant.

**Pull request merged**